Cloudflare Pages - Wrangler fails with `semver` due to Node built-in lib dependency

# Issue summary

While using Wrangler for Cloudflare Pages (`wrangler pages dev` command), with `@shopify/shopify-api` dependency and using them in routes, fails with NodeJS built-in package dependency.

- `@shopify/shopify-api` version: `7.0.0`
- Node version: `v16.19.1`
- Operating system: `macOS`
- Wrangler CLI version: `2.16.0`

```
Compiling worker to "/var/folders/4f/h9qykpcd3pbcbnccyql7bf_m0000gn/T/functionsWorker-0.6486551827953275.mjs"...
✘ [ERROR] 1 error(s) and 0 warning(s) when compiling Worker.


✘ [ERROR] Could not resolve "util"

    ../node_modules/semver/classes/semver.js:19:54:
      19 │ ... TypeError(`Invalid Version: ${require('util').inspect(version)}`)
         ╵                                           ~~~~~~

  The package "util" wasn't found on the file system but is built into node.
  Add "node_compat = true" to your wrangler.toml file to enable Node compatibility.


✘ [ERROR] Build failed with 1 error:

  ../node_modules/semver/classes/semver.js:19:54: ERROR: Could not resolve "util"


error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

## Expected behavior

The `@shopify/shopify-api` package should not depend on NodeJS built-in libs via dependencies.

## Actual behavior

Replace `semver` (which is `node-semver`) with pure JavaScript for consumption in any JavaScript runtime.

## Steps to reproduce the problem

1. Add `@shopify/shopify-api` to Cloudflare Pages project
2. Include `import "@shopify/shopify-api/adapters/cf-worker";`
3. Start `wrangler pages dev` and observe the error

--- a/.changeset/curvy-gorillas-smoke.md
+++ b/.changeset/curvy-gorillas-smoke.md
@@ -2,4 +2,4 @@
 '@shopify/shopify-api': patch
 ---
 
-replace `semver` with `compare-versions` lib to reduce dependency on nodejs builtin-libs
+Replace `semver` with `compare-versions` lib to reduce dependency on nodejs builtin-libs

--- a/.changeset/curvy-gorillas-smoke.md
+++ b/.changeset/curvy-gorillas-smoke.md
@@ -1,0 +1,5 @@
+---
+'@shopify/shopify-api': patch
+---
+
+replace `semver` with `compare-versions` lib to reduce dependency on nodejs builtin-libs

--- a/jest.config.ts
+++ b/jest.config.ts
@@ -1,5 +1,5 @@
 import type {Config} from 'jest';
-import semver from 'semver';
+import {compare as semver} from 'compare-versions';
 
 const projects = [
   './lib/__tests__/jest_projects/library.jest.config.ts',
@@ -11,7 +11,7 @@ const projects = [
 
 // eslint-disable-next-line no-warning-comments
 // TODO Make all projects permanent after support for version 14 is dropped
-if (semver.gte(process.version, '15.0.0')) {
+if (semver(process.version, '15.0.0', '>=')) {
   projects.push(
     './lib/__tests__/jest_projects/adapters.cf-worker.jest.config.ts',
   );

--- a/jest.config.ts
+++ b/jest.config.ts
@@ -1,5 +1,5 @@
 import type {Config} from 'jest';
-import {compare as semver} from 'compare-versions';
+import {compare} from 'compare-versions';
 
 const projects = [
   './lib/__tests__/jest_projects/library.jest.config.ts',
@@ -11,7 +11,7 @@ const projects = [
 
 // eslint-disable-next-line no-warning-comments
 // TODO Make all projects permanent after support for version 14 is dropped
-if (semver(process.version, '15.0.0', '>=')) {
+if (compare(process.version, '15.0.0', '>=')) {
   projects.push(
     './lib/__tests__/jest_projects/adapters.cf-worker.jest.config.ts',
   );

--- a/lib/__tests__/test-helper.ts
+++ b/lib/__tests__/test-helper.ts
@@ -1,5 +1,5 @@
 import * as jose from 'jose';
-import {compare as semver} from 'compare-versions';
+import {compare} from 'compare-versions';
 
 import {shopifyApi, Shopify} from '..';
 import {LATEST_API_VERSION, LogSeverity} from '../types';
@@ -165,7 +165,7 @@ export function testIfLibraryVersionIsAtLeast(
   testFn: jest.ProvidesCallback,
 ) {
   describe(`when library version is at least ${version}`, () => {
-    if (semver(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
+    if (compare(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
       test(testName, testFn);
     } else {
       test.skip(`- '${testName}' requires library version ${version} or higher`, () => {});

--- a/lib/__tests__/test-helper.ts
+++ b/lib/__tests__/test-helper.ts
@@ -1,5 +1,5 @@
 import * as jose from 'jose';
-import semver from 'semver';
+import {compare as semver} from 'compare-versions';
 
 import {shopifyApi, Shopify} from '..';
 import {LATEST_API_VERSION, LogSeverity} from '../types';
@@ -165,7 +165,7 @@ export function testIfLibraryVersionIsAtLeast(
   testFn: jest.ProvidesCallback,
 ) {
   describe(`when library version is at least ${version}`, () => {
-    if (semver.gte(SHOPIFY_API_LIBRARY_VERSION, version)) {
+    if (semver(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
       test(testName, testFn);
     } else {
       test.skip(`- '${testName}' requires library version ${version} or higher`, () => {});

--- a/lib/logger/index.ts
+++ b/lib/logger/index.ts
@@ -1,4 +1,4 @@
-import semver from 'semver';
+import {compare as semver} from 'compare-versions';
 
 import {LogSeverity} from '../types';
 import {ConfigInterface} from '../base-types';
@@ -29,7 +29,7 @@ export type ShopifyLogger = ReturnType<typeof logger>;
 
 function deprecated(logFunction: LoggerFunction) {
   return function (version: string, message: string): void {
-    if (semver.gte(SHOPIFY_API_LIBRARY_VERSION, version)) {
+    if (semver(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
       throw new FeatureDeprecatedError(
         `Feature was deprecated in version ${version}`,
       );

--- a/lib/logger/index.ts
+++ b/lib/logger/index.ts
@@ -1,4 +1,4 @@
-import {compare as semver} from 'compare-versions';
+import {compare} from 'compare-versions';
 
 import {LogSeverity} from '../types';
 import {ConfigInterface} from '../base-types';
@@ -29,7 +29,7 @@ export type ShopifyLogger = ReturnType<typeof logger>;
 
 function deprecated(logFunction: LoggerFunction) {
   return function (version: string, message: string): void {
-    if (semver(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
+    if (compare(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
       throw new FeatureDeprecatedError(
         `Feature was deprecated in version ${version}`,
       );

--- a/lib/setup-jest.ts
+++ b/lib/setup-jest.ts
@@ -1,4 +1,4 @@
-import semver from 'semver';
+import {compare as semver} from 'compare-versions';
 
 import '../adapters/mock';
 import {mockTestRequests} from '../adapters/mock/mock_test_requests';
@@ -131,7 +131,7 @@ expect.extend({
     return {
       message: () =>
         `Found deprecation limited to version ${version}, please update or remove it.`,
-      pass: semver.lt(SHOPIFY_API_LIBRARY_VERSION, version),
+      pass: semver(SHOPIFY_API_LIBRARY_VERSION, version, '<'),
     };
   },
 });

--- a/lib/setup-jest.ts
+++ b/lib/setup-jest.ts
@@ -1,4 +1,4 @@
-import {compare as semver} from 'compare-versions';
+import {compare} from 'compare-versions';
 
 import '../adapters/mock';
 import {mockTestRequests} from '../adapters/mock/mock_test_requests';
@@ -131,7 +131,7 @@ expect.extend({
     return {
       message: () =>
         `Found deprecation limited to version ${version}, please update or remove it.`,
-      pass: semver(SHOPIFY_API_LIBRARY_VERSION, version, '<'),
+      pass: compare(SHOPIFY_API_LIBRARY_VERSION, version, '<'),
     };
   },
 });

--- a/lib/utils/versioned-codeblocks.ts
+++ b/lib/utils/versioned-codeblocks.ts
@@ -1,9 +1,9 @@
-import semver from 'semver';
+import {compare as semver} from 'compare-versions';
 
 import {SHOPIFY_API_LIBRARY_VERSION} from '../version';
 
 export function enableCodeAfterVersion(version: string, fn: () => void): void {
-  if (semver.gte(SHOPIFY_API_LIBRARY_VERSION, version)) {
+  if (semver(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
     fn();
   }
 }

--- a/lib/utils/versioned-codeblocks.ts
+++ b/lib/utils/versioned-codeblocks.ts
@@ -1,9 +1,9 @@
-import {compare as semver} from 'compare-versions';
+import {compare} from 'compare-versions';
 
 import {SHOPIFY_API_LIBRARY_VERSION} from '../version';
 
 export function enableCodeAfterVersion(version: string, fn: () => void): void {
-  if (semver(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
+  if (compare(SHOPIFY_API_LIBRARY_VERSION, version, '>=')) {
     fn();
   }
 }

--- a/package.json
+++ b/package.json
@@ -75,11 +75,9 @@
   },
   "dependencies": {
     "@shopify/network": "^3.2.1",
-    "@types/node-fetch": "^2.5.7",
-    "@types/supertest": "^2.0.10",
+    "compare-versions": "^6.0.0-rc.1",
     "jose": "^4.9.1",
     "node-fetch": "^2.6.1",
-    "semver": "^7.3.8",
     "tslib": "^2.0.3",
     "uuid": "^9.0.0"
   },
@@ -94,7 +92,8 @@
     "@types/express": "^4.17.13",
     "@types/jest": "^29.5.0",
     "@types/node": "^14.18.12",
-    "@types/semver": "^7.3.12",
+    "@types/node-fetch": "^2.5.7",
+    "@types/supertest": "^2.0.10",
     "@types/uuid": "^9.0.0",
     "eslint": "^7.30.0",
     "express": "^4.17.13",

--- a/package.json
+++ b/package.json
@@ -75,7 +75,7 @@
   },
   "dependencies": {
     "@shopify/network": "^3.2.1",
-    "compare-versions": "^6.0.0-rc.1",
+    "compare-versions": "^5.0.3",
     "jose": "^4.9.1",
     "node-fetch": "^2.6.1",
     "tslib": "^2.0.3",

